src/ccs/lcccs.rs

```diff
@@ -0,0 +1,191 @@
+use super::util::{compute_sum_Mz, VirtualPolynomial};
+use super::{CCSShape, CCSWitness};
+use crate::ccs::util::compute_all_sum_Mz_evals;
+use crate::hypercube::BooleanHypercube;
+use crate::spartan::math::Math;
+use crate::spartan::polynomial::MultilinearPolynomial;
+use crate::utils::bit_decompose;
+use crate::{
+  constants::{BN_LIMB_WIDTH, BN_N_LIMBS, NUM_FE_FOR_RO, NUM_HASH_BITS},
+  errors::NovaError,
+  gadgets::{
+    nonnative::{bignat::nat_to_limbs, util::f_to_nat},
+    utils::scalar_as_base,
+  },
+  r1cs::{R1CSInstance, R1CSShape, R1CSWitness, R1CS},
+  traits::{
+    commitment::CommitmentEngineTrait, commitment::CommitmentTrait, AbsorbInROTrait, Group, ROTrait,
+  },
+  utils::*,
+  Commitment, CommitmentKey, CE,
+};
+use bitvec::vec;
+use core::{cmp::max, marker::PhantomData};
+use ff::{Field, PrimeField};
+use flate2::{write::ZlibEncoder, Compression};
+use itertools::concat;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+use sha3::{Digest, Sha3_256};
+use std::ops::{Add, Mul};
+use std::sync::Arc;
+
+/// A type that holds a LCCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct LCCCS<G: Group> {
+  pub(crate) C: Commitment<G>,
+  pub(crate) x: Vec<G::Scalar>,
+  pub(crate) u: G::Scalar,
+  pub(crate) v: Vec<G::Scalar>,
+  // Random evaluation point for the v_i
+  pub r_x: Vec<G::Scalar>,
+  // XXX: This should not need to be here. Should be a reference only.
+  pub(crate) matrix_mles: Vec<MultilinearPolynomial<G::Scalar>>,
+  pub(crate) ccs: CCSShape<G>,
+}
+
+impl<G: Group> LCCCS<G> {
+  // XXX: Double check that this is indeed correct.
+  /// Samples public parameters for the specified number of constraints and variables in an CCS
+  pub fn commitment_key(&self) -> CommitmentKey<G> {
```

Comment:
Would we ever need to sample a new commitment_key for LCCCS, since this type is ever only created starting from a CCCS type?